Only in b/nsis-3.0a2-src/SCons/Tools: crossmingw.pyc
diff -ru a/nsis-3.0a2-src/Source/winchar.cpp b/nsis-3.0a2-src/Source/winchar.cpp
--- a/nsis-3.0a2-src/Source/winchar.cpp	2013-12-08 15:34:38.000000000 +0100
+++ b/nsis-3.0a2-src/Source/winchar.cpp	2016-03-31 17:36:52.705237000 +0200
@@ -94,7 +94,7 @@
   unsigned int v = 0, base = 10, top = '9';
   int sign = 1;
   if (*s == _T('-')) ++s, sign = -1;
-  for ( unsigned int c;; )
+  for ( unsigned short int c;; )
   {
     if ((c = *s++) >= '0' && c <= top) c -= '0'; else break;
     v *= base, v += c;
diff -ru a/nsis-3.0a2-src/Source/writer.cpp b/nsis-3.0a2-src/Source/writer.cpp
--- a/nsis-3.0a2-src/Source/writer.cpp	2013-05-17 21:31:36.000000000 +0200
+++ b/nsis-3.0a2-src/Source/writer.cpp	2016-03-31 17:36:52.705237000 +0200
@@ -57,7 +57,7 @@
   if (m_build_unicode)
   {
     bool strEnd = false;
-    TCHAR ch;
+    TCHAR ch = _T('\0');
     for (; size ; size--)
     {
       if (!strEnd)
